### PR TITLE
Range retries in watermark cursor

### PR DESCRIPTION
**Pull request by daniel-savu**
### Description

Changes the logic in the watermark cursor, to update the range in the `update` method rather than in `get_next_range`. This means that RPC errors won't result in ranges being skipped, so IGP indexing should be more reliable.

### Related issues

- First in a sequence of many, for https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3264

### CODE REVIEW

File: rust/hyperlane-base/src/contract_sync/cursors/mod.rs [hunk 1]
```diff
@@ -242,27 +261,36 @@ where
...
+
+#[async_trait]
+impl<T> ContractSyncCursor<T> for RateLimitedContractSyncCursor<T>
+where
+    T: Send + Debug + 'static,
+{
+    async fn next_action(&mut self) -> Result<(CursorAction, Duration)> {
+        let eta = self.sync_eta()?;
 
         let rate_limit = self.get_rate_limit().await?;
```
Comment (tkporter) [inline rust/hyperlane-base/src/contract_sync/cursors/mod.rs:276]: noting that in this call we update the `tip` and `last_tip_update` regardless of us ever successfully updating the cursor. so if there's an error and we're close to the tip we'll rate limit ourselves and only retry 30s from now

I think this is fine for now but as you continue to refactor we should probably consider this case

File: rust/hyperlane-base/src/contract_sync/cursors/mod.rs [hunk 2]
```diff
@@ -242,27 +261,36 @@ where
...
-        // TODO: Define the sleep time from interval flag
-        Ok((CursorAction::Sleep(Duration::from_secs(5)), eta))
     }
 
     fn latest_queried_block(&self) -> u32 {
         self.sync_state.next_block.saturating_sub(1)
     }
 
-    async fn update(&mut self, _: Vec<(T, LogMeta)>, _range: RangeInclusive<u32>) -> Result<()> {
+    async fn update(&mut self, _: Vec<(T, LogMeta)>, range: RangeInclusive<u32>) -> Result<()> {
```
Comment (tkporter) [inline rust/hyperlane-base/src/contract_sync/cursors/mod.rs:293]: what should we do for Cosmos IGP indexing? Atm `fetch_logs` doesn't propagate any errors that we may get along the way, so we end up calling update here and move onto the next range even if we skipped some blocks https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/dan/watermark-refactoring/rust/chains/hyperlane-cosmos/src/interchain_gas.rs#L231

Open to this being addressed in a different PR but we should try to have a consistent interface between chain log fetching impls when it comes to errors

File: rust/hyperlane-base/src/contract_sync/cursors/mod.rs [hunk 3]
```diff
@@ -80,13 +80,11 @@ impl SyncState {
                 let from = self.next_block;
                 let mut to = from + self.chunk_size;
                 to = u32::min(to, tip);
-                self.next_block = to + 1;
```
[the remark this answers is not in the document]
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/mod.rs:83]: good point, changed

File: rust/hyperlane-base/src/contract_sync/cursors/mod.rs (hunk 2 above)
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/mod.rs:293]: pushed a fix for this, lmk if it lgty

File: rust/hyperlane-base/src/contract_sync/cursors/mod.rs (hunk 1 above)
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/mod.rs:276]: I think this was the case before right? Less critical than other bits but I took a note

**Pull request merged**